[Feature] Support group switching with v0 UI

Will-Howard opened the pull request:
# Description

Summary:
- New endpoint `group-switching/available.ts` to fetch the groups and discussions that are available to switch into
- New endpoint `group-switching/index.ts` for submitting switch requests
- The modal itself, `GroupSwitchModal`
- A bunch of schema changes to support this

These are the [designs](https://www.figma.com/design/6PJJZThivMuDTZ9xcoWvBs/Group-switching-UI?node-id=3066-1891&t=U7nnbjPTr0OHDICS-0). This PR doesn't match them that closely, I'm optimising for having a usable version with more basic styling out first, and will then come back and make the UI match the designs exactly

### Testing pointers

I've been testing with my own user against the "What the fish?" test course. To do this yourself (not 100% sure this is correct):
1. Visit the course page (`/courses/fish-test-course/1`) to ensure there is a "Course registration" for this course and your user
2. Add this "Course registration" as a participant in one of the [groups](https://airtable.com/appPs3sb9BrYZN69z/tblyiJSPoniwhi17T) for this course
3. Add this "Course registration" as a participant (expected) in one of the [group discussions](https://airtable.com/appPs3sb9BrYZN69z/tblDNME0bA9OoApTk) for this course

## Issue
<!-- If this PR is related to a project, and there's no related issue, link this PR to the project -->

#1210 

## Developer checklist

- [x] Front-end code follows [Component Accessibility Checklist (for Testing)](https://github.com/bluedotimpact/bluedot/blob/master/DEVELOPMENT_HANDBOOK.md#component-accessibility-checklist)
- [ ] Considered having snapshot tests and/or happy path functional tests
- [ ] Considered adding Storybook stories

<!-- You might also want to check the tests locally with `npm run test`, although CI will check this for you -->

## Screenshot

<!-- If this PR results in visual changes -->

| 📸 |  |
|---------|---|
| 🖥️ | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 25 23" src="https://github.com/user-attachments/assets/6c2e2cc8-de54-41fe-ae6e-a7ee657d635e" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 25 38" src="https://github.com/user-attachments/assets/30de2fe5-a8c2-4113-b6ff-ed859ce40d2e" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 25 43" src="https://github.com/user-attachments/assets/a8b9b867-5b14-4245-9e25-78af43be5bc6" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 34 03" src="https://github.com/user-attachments/assets/8a8399e3-f21d-4c95-a9ca-00c43669845e" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 34 25" src="https://github.com/user-attachments/assets/da8d17c4-8b27-488c-be20-9e3b3fbdea3a" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 35 19" src="https://github.com/user-attachments/assets/d686a261-492d-48d1-95f6-ed73edfebe4d" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 35 26" src="https://github.com/user-attachments/assets/9eaa2b90-ee28-4e39-a061-2bf5409ebd52" /> |
| | <img width="2086" height="1634" alt="Screenshot 2025-08-25 at 19 35 51" src="https://github.com/user-attachments/assets/f02e2f96-969d-4032-8494-cfdeaff0ced9" /> |
| 📱  | <img width="662" height="1458" alt="Screenshot 2025-08-25 at 19 37 24" src="https://github.com/user-attachments/assets/9aca2d02-7b22-4d5a-a680-376b81c92f73" /> |
| | Otherwise as above, the mobile styles currently don't look great but the modal is usable |

Will-Howard commented on libraries/ui/src/Modal.tsx:
```diff
@@ -25,7 +25,7 @@ export const Modal: React.FC<ModalProps> = ({
       isOpen={isOpen}
       onOpenChange={setIsOpen}
-      className="fixed inset-0 z-10 overflow-y-auto bg-black/25 flex min-h-full items-center justify-center p-4 backdrop-blur-xs"
+      className="fixed inset-0 z-60 overflow-y-auto bg-black/25 flex min-h-full items-center justify-center p-4 backdrop-blur-xs"
```
Previously the header would appear on top of this on small screens. I assume this wasn't intentional

Will-Howard commented on apps/website/src/components/courses/GroupDiscussionBanner.tsx:
```diff
@@ -103,17 +116,364 @@ const GroupDiscussionBanner: React.FC<GroupDiscussionBannerProps> = ({
+  // people in multiple groups too. We're not explicitly supporting that case at the moment, but
+  // we do at least display the group/discussion the user is switching out of so that
+  // they can notice and request manual support.
+  const oldGroup = groups.filter((g) => g.userIsParticipant)[0];
+  const oldDiscussion = discussions.filter((d) => d.userIsParticipant)[0];
```
(This will fail validation on the back end if these are ever undefined)

marn-in-prod commented on apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts:
```diff
@@ -0,0 +1,231 @@
+  const roundId = participant.round;
+
+  const round = await db.get(roundTable, { id: roundId });
+
+  /**
+   * Get groups the user is allowed to switch to.
+   *
+   * KNOWN ISSUE: Ideally this would get the buckets that the participant is a member of,
+   * and then look up groups that allow that bucket. Currently the Buckets table in the
+   * course runner base only has the People in the bucket as a list of names like "John Doe | AI Alignment (2024 Mar)",
+   * so we can't reliably look up by participant id.
+   *
+   * WORKAROUND: Look up the group (should be one group, but support many) the user is currently in, then look up
+   * groups that share a bucket with that group. Assume that if the user is allowed in their current group, they
+   * will be allowed in these groups that share a bucket.
+   */
```
Similar issues are spread across all of our endpoints because we can't verify uniquness. I think just having a single generic github issue and linking to that issue with maybe a single line explaining this particular edge case is good enough. 

For example, users can have multiple course registrations for the same MOOC on profile pages, it's possible to create multiple users with the same email address etc.

marn-in-prod commented on apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts:
I commented on this file in a different PR so keep that in mind when merging everything

marn-in-prod reviewed (approved):
In general, I think look at all of the comments across the PRs before you start to merge. 

I approved since I think getting this out is important but wanted to still give you the direction that I think would make this easier to follow in the future.

Will-Howard commented:
I'm about to merge this now, summary of what I changed based on the review comments:
1. I removed `notNull()` constraints for all added fields (apart from arrays, which fall back to being empty already)
2. I removed the long comment flagged in [this comment](https://github.com/bluedotimpact/bluedot/pull/1233#discussion_r2298956664) and created a github issue for it (#1249)
3. I updated the logic in `available.ts` to be more readable (based on [this comment](https://github.com/bluedotimpact/bluedot/pull/1233#discussion_r2298950603))
